### ISSUE TITLE
Enable DBBC3 continous amplitude calibration ("tsys monitoring") with FS10

### ISSUE THREAD

**Issue by varenius** [nvi-inc/fs#32]
For VGOS observations using DBBC3, and in particular for source-structure interpretation of the correlated data, we need (continuous) logging of tsys values to a log file during observations.

@wehimwich informed me via email 2019-10-01 regarding this that: 
> Continuous cal worked, but the communication required overloaded the system. It will have to wait until [DBBC3] multicast is available.

Yesterday I read in an email from Gino Tuccari (DBBC3 developer) that

> [multicast is] available the v124 V version, which produces multicast strings and it's just like the version v123 in terms of firmware.  [...] I don't think it's compatible one-to-one with the RDBE multicast, because we developed independently, but I suppose this could in some way compared or adapted for the FS.
 
I have asked Gino if there is an interface document etc. to describe the structure of the multicast. Hopefully the community can work together to adapt the FS10 to parse the DBBC3 multicast in a way similar to the RDBE.

**Comment by varenius** [comment]
Email from DBBC3 developer Sven Dornbush yesterday:

> here is the description of the DBBC3 Multicast Format for DDC Mode.
> Most of the data is included in both the DDC_V v124 and in the DDC_U v125 versions.
> If data is only included in the DDC_U (VDIF timestamp f.e.) there is a note in the documentation.
> 
> The DDC_V v124 from the Hatlab-Homepage also includes a simple Python Parser that parses the Multicast-Packets and gives out the content in the console.
> This can be used as reference or for verifying your own modifications to the FS.
> 
> The DDC_V v124 uses the same Firmware as the DDC_V v123 you are already using for VGOS observations, it is just and extension providing multicast support.
> The main config file has two additional lines at the end, one for the IP Address of the DBBC3, one for the Multicast-Group. The IP Address needs to be correct,
> otherwise the Control Software could throw an Exception.

[DBBC3 DDC Multicast Format Description.pdf](https://github.com/nvi-inc/fs/files/4508081/DBBC3.DDC.Multicast.Format.Description.pdf)

I guess we can get the total power values, for each BBC, from the DBBC3 at the bytes described by these lines in the PDF BBC-information table:
```
Total Power Value for BBC001 USB (contcal status On)
Total Power Value for BBC001 LSB (contcal status On)
Total Power Value for BBC001 USB (contcal status Off)
Total Power Value for BBC001 LSB (contcal status Off)
```
This should give the Total Power Indicator values "TPIcal" and "TPInocal" for each spectral window (BBC USB/LSB channel). Using the mentioned python-script to extract the TPI values, and calibration diode level Tcal [K], one should be able to calculate Tsys (as proxy for SEFD) continuously during the experiment as:

Tsys = Tcal * TPInocal / (TPIcal-TPInocal)

Using this proxy Tsys value, together with known antenna Gain and DPFU [K/Jy] vs frequency, one can obtain the SEFD values needed to scale the correlated visibility amplitudes to Jansky.

I will try this once the Onsala VGOS antennas are operational again (LNAs currently being repaired).

**Comment by varenius** [comment]
I have now tested extracting TPI values from the DBBC3 DDC_V 124 multicast data. It works! This means we can now do total power monitoring with the DBBC3! This is outside the field-system, and it requires some manual work to actually get the values into a useful data format for end-user data calibration. But all the information needed should be present. I have some questions on details (e.g. it seems the on/off TPI channels are swapped with respect to documentation) which I will iterate with Gino/Sven. But once I have something fairly robust, I will upload the script and a description here. (Basically the parser.py distributed in the 124-firmware file works; but I had to change to specifying a specific network interface instead of "any" to get any packets. Also one can (of course) format the output in a more user-friendly way).

Having a work-around using an outside multicast-monitor script solves the immediate problem of getting the total power data every second. Long-term, it would perhaps be advantageous if this was integrated into the FS itself. We can perhaps discuss this after we've sorted out any remaining questions with this python-approach.

But it's great to see that we can indeed extract TPI values!

**Comment by varenius** [comment]
In case anyone is interested, here is my current receiver script:
https://github.com/varenius/oso/blob/master/VGOS/DBBC3-124-multicast-parser.py
which is just a tiny bit modified from the script supplied in the 124-firmware package.

**Comment by varenius** [comment]
Some notes from recent investigations, so far on FS 9.12 (waiting on software engineer to apply various improvements to our ST-library before FS10 testing):
- It seems cont_cal=on works for fivept and ONOFF with DBBC3 (need to be verified with FS10). This is good because it means we don't have to switch between cont_cal=off for these tasks, and cont_cal=on for the multicast-logging (which needs to have cont_cal on for the DBBC3 to be able to continuously track the antenna power).
- As indicated by previous emails, using tpicd with cont_cal=on creates problems. I haven't tested all aspects, but I noted right away that trying to send configuration to DBBC3 (e.g. bbc001=...) becomes un bearably slow if running tpicd with cont_cal at the same time. It seems the amount of data sent by DBBC3 to tpicd requires all available bandwidth on the communication channel. So, moving to multicast-reception of the TPI values seems like a good option.
- I have now tested receiving multicast packets with TPI values (and other things) from DBBC3 for a full 24h VGOS experiment (VO0132 for antenna Oe). In principle this works nicely! In practice, there are some minor tweaks that should be implemented (such as dbbc3 cont_cal display_swap not being applied to the multicast stream, causing the Oe TPI data - which uses cont_cal=2,80 - to be negative, i.e. TpiOff>TpiOn --> negative SEFD). But it works. 
- Logging all (useful for this firmware version; not dummy lines) DBBC3 multicast information, emitted once per second, to a silly ASCII file without any clever optimisations seem to result in about 1MB of data per minute, or 1.2GB for a 24h experiment. This can then be reduced further by parsing the output-file and a) only using values within scans (the multicast is emitted continuosly, also when we are not recording), b) extracting only the TPI-data, and c) averaging the TPI data to e.g. one value every 5 seconds. This results in an ASCII file of size 2.7MB for a full 24h observation, logging all BBCs separately.
- I attach a screen-shot of 2h of data from VO0132 OE where I have done the above, i.e. extracting SEFD vs time averaged to 5 seconds for when we are recording. There are some huge outliers (as is the case with TPI-data in ordinary FS logfiles from e.g. S/X systems) excluded by the plot ranges. The outliers could be removed by an antabfs-like script, which is easy to develop once we settle on the data formats.
- DBBC3 developers will likely change the multicast format slightly in the future, to allow for some useful changes (like including the full BBC frequency and not only the integer part). After this has settled, we should be able to settle on a production way of doing this.
- For reference: The way I obtained these values were:
1. cont_cal=on for FS
2. configure DBBC3 and FS for ordinary VGOS frequency setup
3. run ONOFF ten times or so on Cas A to get the Noise Diode strength in Jansky
4. parse the ONOFF FS log file and average the diode-values together and get a table of frequency and diode reference-level in Jy. (This can perhaps be done better by ACQUIR and GNPLT, and using the RXG file, but I was lazy and did this with python).
5. Receive the DBBC3 multicast from 124 firmware, every second. In addition to storing the TPI values, I on-the-fly calculate the SEFD using the diode-level in Jansky from the TPI values and write this to a logfile. Total ca 1.2GB for 24h. 
6. After experiment (could also be done on-the-fly): parse FS logfile to find scan times from data_valid=on/off. Extract the Multicast values for the scan-ranges, and average them together for 1 value every 5 seconds. Store this to a new logfile, total ca 2.7MB for 24h. Note: A silly for-loop heavy python-program to parse the 1.2GB of data finished in just a few seconds. This could easily be optimized in many ways, but even this silly approach appears to be fine.
7. Plot the result. Hard to see structure when looking at all 24h, so here is 2h as an example:
<img width="1233" alt="VO0132_OE_SEFD_vs_time_2h" src="https://user-images.githubusercontent.com/8872291/81776110-7d16e080-94ee-11ea-819e-283c18784146.png">
* A comment on the figure: This does not show the dependence on pointing direction, but it is very clear when comparing the SEFD with source altitude that observations closer to the horizon yield a much higher SEFD compared to observations at a higher angle from the ground, in line with expectations. 

* Note about SEFD/Tsys: SEFD is what one wants to calibrate the obtained correlation coefficients Cij on baseline i-i to get calibrated "visibilities" in Janskys for astronomical sources:
Vij = Cij * sqrt(Ni*Nj) / β
where Vij is the visibility amplitude in Jy, β is a dimensionless factor taking into account the effects of digitization, and Ni and Nj represent the system noise of the two antennas expressed as a Source Equivalent Flux Density (SEFD) in Jy. One could also log Tsys, and translate to SEFD using some assumed or measured Degrees Per Flux Unit (DPFU) value (and gain curve), but I think it is simpler to just do things in SEFD all the way.

* Re FS integration: It should be possible to get the FS to receive the multicast and store the relevant values in the FS-log. If only storing the scan times, it should just be a few MB extra (max 15MB extra if using 1 value every second) for a 24h observation. From there, one could use an antabfs-like tool to read and export the data into a format (e.g. antab files) for further analysis.

* Final note: I can share the scripts I used for the various steps above, but at this point they are a bit ugly and as the DBBC3 format may change slightly, I won't advertise them for now. But it should be straight-forward to obtain

**Comment by wehimwich** [comment]
Hi Eskil,

This looks great. It looks like using the multicast for Tsys data is very promising. This is the first verification that I have seen of this. It is very helpful.

On FS integration, there are a few issues. As I think you the know (and if my memory is correct), the _antabfs_ program works with recorded values in the log. I have forgotten whether it uses the raw count data or the uses the calculated Tsys values by backing out the cal values. In any event, the (possibly updated) `.rxg` files are then used to calculate the correct Tsys values for the `antab` file. In principle,  two things are needed:

- [ ] Get the multicast values into the log in the correct format

- [ ] Modify _antabfs_ to handle the DBBC3 cont-cal data,  which should be very similar to what is done for the DBBC2 cont-cal data.

We have a development effort for a C program that integrates into the FS, using its internal infrastructure, for the first item. This will ultimately use the Tcal values from the `.rxg` files to calculate the Tsys values at acquisition time. This will integrate with a Tsys display window in the FS well. The second item is not covered, but presumably there would be a plan in the EVN for this.

I have a couple questions. I am hoping you know the answers, or maybe there is an expert close by to you (Michael?) that does:

1. Does _antabfs_ use the acquisition value of the Tcal by channel or does it work with the raw counts? I know _gnplt_ uses the Tcal value in the log, so we will provide that in any event.

2. Does _antabfs_ get the channel set-up from the schedule (VEX) file or from the FS log? I would guess that it is the schedule. That would make it pretty much backend independent for the setup information.

Thanks, Ed

**Comment by varenius** [comment]
Hi Ed,
1) As far as I remember, antabfs uses the TPI on/off values, plus the tcal, and gain from the RXG file. 
2. antabfs deduces the sky frequencies and setup from the FS log, at least last time I used it. 
You can check yourself, as the latest version of antabfs should be on github: 
https://github.com/evn-vlbi/VLBI-utilities/blob/master/antabfs.py

Extra note: As per my DBBC3 testing, I was told that Sven is working on a revised DBBC3 multicast which will adress some minor suggestions I had. It should be out any day now, so perhaps worth waiting for that before finalising the format within the FS?

**Comment by varenius** [comment]
Note: This has now been fixed, and is working with the ed/dbbc3 branch. But we should perhaps leave this issue open until it has been merged into the main branch.

**Comment by wehimwich** [comment]
I think the FS side of the task list has been closed for sometime, at least as far back as FS _10.1.0_ e135ce6b48970c0f3fb9b43c2a719afe8552f8d6:

> * [x]  Get the multicast values into the log in the correct format
> * [ ]  Modify _antabfs_ to handle the DBBC3 cont-cal data,  which should be very similar to what is done for the DBBC2 cont-cal data.

Since _antabfs_ is not part of the FS, I am going to close this issue. Maybe someone knows the status of _antabfs_ and can comment on that.
